pkg/cache/resource.go

```diff
@@ -76,3 +72,46 @@ func GetResourceName(xds proto.Message) string {
 		return ""
 	}
 }
+
+// GetResourceReferences returns the names for dependent resources (EDS cluster
+// names for CDS, RDS routes names for LDS).
+func GetResourceReferences(resources ...Resource) map[string]bool {
+	out := make(map[string]bool)
+	for _, res := range resources {
+		if res == nil {
+			continue
+		}
+		switch v := res.(type) {
+		case *v2.ClusterLoadAssignment:
+			// no dependencies
+		case *v2.Cluster:
+			// for EDS type, use cluster name or ServiceName override
+			if v.Type == v2.Cluster_EDS {
+				if v.EdsClusterConfig != nil && v.EdsClusterConfig.ServiceName != "" {
+					out[v.EdsClusterConfig.ServiceName] = true
+				} else {
+					out[v.Name] = true
+				}
+			}
+		case *v2.RouteConfiguration:
+			// no dependencies
```

Comment:
why? Route depends on the clusters to work properly..

Reply:
It does not depend in the ADS sense. Envoy queries the entire set of clusters, not individual ones by name. It's valid to send a route with no cluster, or to send more clusters than used.

Reply:
Good point though -- expanded the comment in the code.